### PR TITLE
Support mutual TLS for registry

### PR DESCRIPTION
**Pull request by johntdyer**

### CODE REVIEW

**Comment by johntdyer** [comment]
@wvanbergen  Any thoughts on merging this ?

File: lib/schema_registry/client.rb [hunk 1]
```diff
@@ -27,11 +28,21 @@ def initialize(code, message)
 
   class Client
 
-    attr_reader :endpoint, :username, :password
+    attr_reader :endpoint, :username, :password, :client_key, :client_certificate, :ca_certificate, :verify_mode
 
-    def initialize(endpoint, username = nil, password = nil)
+    def initialize(endpoint, username = nil, password = nil, ssl_options = {})
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: Can you use keyword arguments here, rather than a hash?

File: lib/schema_registry/client.rb [hunk 2]
```diff
@@ -100,6 +157,6 @@ def request(method, path, body = nil)
           raise error_class.new(response_data['error_code'], response_data['message'])
         end
       end
-    end
-  end
-end
+    end # request
+  end # Client
+end # SchemaRegistry
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: Not a big fan of these comments, can you :fire: them?

File: lib/schema_registry/version.rb [hunk 3]
```diff
@@ -1,3 +1,3 @@
 module SchemaRegistry
-  VERSION = "0.0.4"
+  VERSION = "0.1.0"
```
Comment (wvanbergen) [inline lib/schema_registry/version.rb]: I rather do this in a separate commit that does the release.

**Review by wvanbergen — CHANGES_REQUESTED** [review]
Thanks for your contribution! I think this makes sense to support. I just some small remarks about the implementation to address before merging.

File: lib/schema_registry/version.rb (hunk 3 above)
Comment (johntdyer) [reply lib/schema_registry/version.rb]: fixed

File: lib/schema_registry/client.rb (hunk 2 above)
Comment (johntdyer) [reply lib/schema_registry/client.rb]: removed

File: lib/schema_registry/client.rb (hunk 1 above)
Comment (johntdyer) [reply lib/schema_registry/client.rb]: @wvanbergen  mind if I ask why ?

Comment (johntdyer) [reply lib/schema_registry/client.rb]: done

**Comment by johntdyer** [comment]
@wvanbergen  - Fixed

File: lib/schema_registry/client.rb [hunk 4]
```diff
@@ -27,11 +28,21 @@ def initialize(code, message)
 
   class Client
 
-    attr_reader :endpoint, :username, :password
+    attr_reader :endpoint, :username, :password, :client_key, :client_certificate, :ca_certificate, :verify_mode
 
-    def initialize(endpoint, username = nil, password = nil)
+    def initialize(endpoint, username = nil, password = nil, client_certificate = nil, client_key = nil, ca_certificate = nil, verify_mode = :verify_peer)
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: Whoops, I should have been more clear. I was talking about proper keyword arguments, no optional arguments.

``` ruby
  def initialize(endpoint, username = nil, password = nil, client_certificate: nil, client_key: nil, ca_certificate: nil, verify_mode: OpenSSL::SSL::VERIFY_PEER)
```

File: lib/schema_registry/client.rb [hunk 5]
```diff
@@ -27,11 +28,21 @@ def initialize(code, message)
...
       @endpoint = URI(endpoint)
       @username, @password = username, password
+      
+      unless client_certificate.nil? && client_key.nil?
+        @client_certificate = client_certificate
+        @client_key = client_key
+        @ca_certificate = ca_certificate
+        @verify_mode = OpenSSL::SSL.const_get(
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: I prefer using the `OpenSSL::SSL` constants directly, rather than symbols.

File: lib/schema_registry/client.rb [hunk 6]
```diff
@@ -55,8 +66,54 @@ def default_compatibility_level=(level)
...
+
+      if @client_key != nil
+        if File.exists?(@client_key)
+          options.merge!(
+            key: OpenSSL::PKey::RSA.new(File.read(@client_key))
+          )
+        else
+          raise ArgumentError, "client key file not found [#{@client_key}]"
+        end
+      end # client_key
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: Style nitpick: more comments after `end`

File: lib/schema_registry/client.rb [hunk 7]
```diff
@@ -55,8 +66,54 @@ def default_compatibility_level=(level)
...
+          options.merge!(
+            cert: OpenSSL::X509::Certificate.new(File.read(@client_certificate))
+          )
+        else
+          raise ArgumentError, "client cert file not found [#{@client_certificate}]"
+        end
+      end #client_cert
+
+      return options
+    end # build_ssl_options
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: Style nitpick: more comments after `end`

File: lib/schema_registry/client.rb [hunk 8]
```diff
@@ -55,8 +66,54 @@ def default_compatibility_level=(level)
       request(:put, "/config", compatibility: level)
     end
 
+    # Build options hash for net/http based on params provided. Primary for selectivly adding TLS config options for MTLS
+    def build_ssl_options
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: What is the value add of this method? What if we would ask for an argument, that we simply forward to `Net::HTTP` as is, rather than constructing our own? `Net::HTTP` can be a bit awkward, but it's very well documented so it should be relatively easy for people to figure things out.

Comment (wvanbergen) [reply lib/schema_registry/client.rb]: We could always add a helper function if we consider "raw" `Net::HTTP` to cumbersome.

``` ruby
# raw
Client.new('https://foo.bar', use_ssl: true, verify_mode: OpenSSL::SSL::VERIFY_PEER)

# with helper
Client.new('https://foo.bar', Client.connection_options(...))
```

File: README.md [hunk 9]
```diff
@@ -23,6 +23,22 @@ Or install it yourself as:
 
 TODO: Write usage instructions here
 
+
+### TLS
+
+Support for MTLS
```
Comment (wvanbergen) [inline README.md]: *M*TLS?

**Review by wvanbergen — CHANGES_REQUESTED** [review]

File: lib/schema_registry/client.rb (hunk 5 above)
Comment (johntdyer) [reply lib/schema_registry/client.rb]: k, but I am using this from another module ( logstash ) so its harder to pass an object from the logstash conf to the logstash codec and then to your module

File: lib/schema_registry/client.rb (hunk 8 above)
Comment (johntdyer) [reply lib/schema_registry/client.rb]: @wvanbergen  I'd be happy w/ the helper method approach but I am not sure how to go about it based solely on your example

Comment (wvanbergen) [reply lib/schema_registry/client.rb]: Something like this:

1. Make the client class as straightforward as possible, simply forwarding stuff to Net::HTTP
2. Create a helper method that makes it easier to instantiate such a class.

E.g. the class would look something like this:

``` ruby
class Client
  def initialize(endpoint, username = nil, password = nil, connection_options = {}
    ...
    @connection_options = connection_options
    ...
  end

  def request
    ...
    Net::HTTP.start(endpoint.host, endpoint.port, @connection_options) do |http| ... end
  end
end
```

For the helper method, you could either make a static method that returns a hash that you could use for the `connection_options` argument. Or you could create additional constructors.

``` ruby
class Client
  def self.connection_options(...)
    ...
    return { ... }
  end
end

Client.new('https://foo.bar', 'user', 'pass', Client.connection_options(...))
```

or, as an additional constructor:

``` ruby
class Client
  def self.from_config(configuration)
    ...
    new(endpoint, username password, connection_options_from_config)
  end
end
```

I'm fine with either option, so whatever you think is best. As long as the user is able to simply provide options that will be passed to `Net::HTTP`, and can set any option as they see fit, our API will never be limiting them.

**Comment by johntdyer** [comment]
@wvanbergen  - I went w/ the splat operator and a static method... What do you think?

File: lib/schema_registry/client.rb [hunk 10]
```diff
@@ -48,28 +50,69 @@ def subject(name)
...
+
+      options
     end
 
     def request(method, path, body = nil)
-      Net::HTTP.start(endpoint.host, endpoint.port, use_ssl: endpoint.scheme == 'https') do |http|
+      # build config for http client
+      default_options = {
+        use_ssl: endpoint.scheme == 'https',
+        verify_mode: @verify_mode
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: I think this is a leftover from a previous iteration.

Comment (johntdyer) [reply lib/schema_registry/client.rb]: 🤦‍♂️

Comment (johntdyer) [reply lib/schema_registry/client.rb]: pushed

Comment (johntdyer) [reply lib/schema_registry/client.rb]: @wvanbergen  thanks, just pushed a fix

File: README.md [hunk 11]
```diff
@@ -23,6 +23,45 @@ Or install it yourself as:
...
+
+You have two options of passing options to net/http.  You can pass in all the config as named parameters
+
+##### Example
+
+```ruby
+    @client = SchemaRegistry::Client.new(
+        "https://testschmreg.domain.com:8082",
+        '',
+        '',
```
Comment (wvanbergen) [inline README.md]: Note: you can leave those two arguments out because they are optional:

``` ruby
@client = SchemaRegistry::Client.new(
  "https://testschmreg.domain.com:8082",
  cert: ...,
  key: ...,
)
```

They will be set to `nil`

File: README.md [hunk 12]
```diff
@@ -23,6 +23,45 @@ Or install it yourself as:
...
+
+The second option is to use a helper static method to avoid having to handle the encoding.
+
+##### Helper Example
+
+```ruby
+    @client = SchemaRegistry::Client.new(
+        "https://testschmreg.domain.com:8082",
+      '',
+      '',
```
Comment (wvanbergen) [inline README.md]: Same here

File: lib/schema_registry/client.rb [hunk 13]
```diff
@@ -15,31 +16,32 @@ def initialize(code, message)
...
+    40_402 => (VersionNotFound           = Class.new(SchemaRegistry::ResponseError)),
+    40_403 => (SchemaNotFound            = Class.new(SchemaRegistry::ResponseError)),
+    42_201 => (InvalidAvroSchema         = Class.new(SchemaRegistry::ResponseError)),
+    42_202 => (InvalidVersion            = Class.new(SchemaRegistry::ResponseError)),
+    42_203 => (InvalidCompatibilityLevel = Class.new(SchemaRegistry::ResponseError)),
     409   => (IncompatibleAvroSchema    = Class.new(SchemaRegistry::ResponseError)),
-    403   => (UnauthorizedRequest       = Class.new(SchemaRegistry::ResponseError)),
-  }
+    403   => (UnauthorizedRequest       = Class.new(SchemaRegistry::ResponseError))
+  }.freeze
```
Comment (wvanbergen) [inline lib/schema_registry/client.rb]: I am seeing a bunch of unrelated style changes (also single vs. double quotes, indentation, etc). I prefer to leave these out of scope for this PR so that this PR is only about adding TLS support.

If you feel strongly about them, you can open a separate PR to get these things fixed.

**Review by wvanbergen — APPROVED** [review]
Couple of final remarks but we're close to the finish line. :)

File: README.md (hunk 12 above)
Comment (johntdyer) [reply README.md]: @wvanbergen  pushed

File: lib/schema_registry/client.rb (hunk 13 above)
Comment (johntdyer) [reply lib/schema_registry/client.rb]: yea, I ran it though rubocop... I'l revert

**Comment by wvanbergen** [comment]
:tada: this looks great now. Thanks for your patience while working with me here.

Final question: can you squash some of the commits? 15 commits for this is a bit over the top. I'll merge afterwards

**Comment by johntdyer** [comment]
@wvanbergen  squashed , thanks !

**Comment by johntdyer** [comment]
@wvanbergen  - Any idea when you might cut a release to rubygems for this ?

**Comment by wvanbergen** [comment]
Already done!

**Comment by johntdyer** [comment]
nm, I see it :) Thanks!

**Comment by johntdyer** [comment]
awesome, thanks !
